File: submissions/0002-add-two-numbers/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        if(l1 == null) {
            return l2;
        }
        
        ListNode result = l1;
        ListNode l1Last = null;
        int carry = 0;
        while(l1 != null && l2 != null) {
            int sum = l1.val + l2.val + carry;
            l1.val = sum % 10;
            carry = sum / 10;
            l1Last = l1;
            l1 = l1.next;
            l2 = l2.next;
        }
        
        // some of l1, l2 are null;
        if(l1 == null) {
            l1Last.next = l2;
        }
        
        while(carry != 0 && l1Last.next != null) {
            int sum = l1Last.next.val + carry;
            l1Last.next.val = sum % 10;
            carry = sum / 10;
            l1Last = l1Last.next;
        }
        
        if(carry!=0) {
            l1Last.next = new ListNode(carry);
        }
        
        return result;    
    }
    
}

File: submissions/0002-add-two-numbers/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        
        result_holder = ListNode()
        result_curr = result_holder

        carry = 0

        while l1 and l2:
            val = l1.val + l2.val + carry
            carry = val // 10
            val = val % 10
            l1.val = val

            result_curr.next = l1
            result_curr = l1
            l1 = l1.next
            l2 = l2.next

        remain = None
        if l1:
            remain = l1

        if l2:
            remain = l2

        while remain:
            result_curr.next = remain
            if carry == 0:
                break
            val = remain.val + carry
            carry = val // 10
            val = val % 10
            remain.val = val
            result_curr = remain
            remain = remain.next
        
        if carry == 1:
            node = ListNode(1)
            result_curr.next = node


        return result_holder.next

        
